modin/experimental/core/execution/native/implementations/hdk_on_native/dataframe/dataframe.py

```diff
@@ -1038,350 +1071,107 @@ def _union_all(
         HdkOnNativeDataframe
             The new frame.
         """
-        frames = [self] + other_modin_frames
-
-        # This is a special case, where we need to preserve the index of empty frames.
-        if any(len(f.columns) == 0 for f in frames):
-            return self._union_all_arrow(frames, join, sort, ignore_index)
+        index_cols = None
+        columns = OrderedDict()
+        for col in self.columns:
+            columns[col] = self._dtypes[col]
 
-        # determine output columns
-        def join_cols():
-            new_cols_map = OrderedDict()
-            for col in self.columns:
-                new_cols_map[col] = self._dtypes[col]
+        if join == "inner":
             for frame in other_modin_frames:
-                if join == "inner":
-                    for col in list(new_cols_map):
-                        if col not in frame.columns:
-                            del new_cols_map[col]
-                elif join == "outer":
-                    for col in frame.columns:
-                        if col not in new_cols_map:
-                            new_cols_map[col] = frame._dtypes[col]
+                for col in list(columns):
+                    if col not in frame.columns:
+                        del columns[col]
+        elif join == "outer":
+            for frame in other_modin_frames:
+                for col in frame.columns:
+                    if col not in columns:
+                        columns[col] = frame._dtypes[col]
+        else:
+            raise NotImplementedError(f"Unsupported join type {join=}")
+
+        frames = []
+        for frame in [self] + other_modin_frames:
+            # Empty frames are filtered out only in case of the outer join.
```

Comment:
Has there been filtering for empty frames in this function before, and if so, where?

Reply:
Here - [_union_all_arrow()](https://github.com/modin-project/modin/pull/6251/files#diff-d1a0db76b871ca475d604ff109cc3ebf13d7ead8ce07cca7caa54ace5f736cc0L1256)